File: lib/svg/elements/utils/mathHelpers.py
from math import cos as COS
from math import floor, pi, sqrt
from math import log as ln
from math import radians as rad
from math import sin as SIN
from math import tan as TAN
from random import randint as RANDN
from random import random as RAND


def my_round(x, p):
    return round(floor(x * 10**p) / 10**p, ndigits=3)


class Rounder:
    def __init__(self, precision=0):
        self.dx = 0
        self.dy = 0
        self.precision = precision

    def __call__(self, x: float, y: float) -> tuple[int, int]:
        newX = my_round(self.dx + x, self.precision)
        newY = my_round(self.dy + y, self.precision)
        self.dx += x - newX
        self.dy += y - newY

        return newX, newY


def sin(angle):
    return SIN(rad(angle))


def cos(angle):
    return COS(rad(angle))


def norm2(v):
    return (v[0] ** 2 + v[1] ** 2) ** 0.5


def dist2(a, b):
    return norm2((a[0] - b[0], a[1] - b[1]))


def normInf(v):
    return max(abs(v[0]), abs(v[1]))


def distInf(a, b):
    return normInf((a[0] - b[0], a[1] - b[1]))


def randint(a, b=None, len=1):
    if b is None:
        b = a
        a = 0

    if len > 1:
        return (RANDN(a, b) for _ in range(len))
    else:
        return RANDN(a, b)


a = 2 * pi / 3
I = (3 / pi) * ln(2 + sqrt(3)) - 0.96
K = 1 / I


def f(x):  # distribution
    return K * (1 / (a * (x - 0.5)) - 0.96)


def F(x):  # repartition
    u = a * (x - 0.5)
    return K * ((1 / a) * (ln(1 / COS(u) + TAN(u)) + ln(2 + sqrt(3))) - 0.96 * x)


def random():
    u = RAND()
    x = u
    eps = 1e-9
    for _ in range(6):
        if abs(x - 0.5) < eps:
            x += eps

        fx = f(x)
        if fx != 0:
            x -= (F(x) - u) / fx

        x = min(max(x, eps), 1 - eps)
    return x

File: lib/svg/elements/utils/mathHelpers.cc
#include "mathHelpers.hh"

#include <cmath>
#include <random>
#include <algorithm>

const double a = 2.0 * std::numbers::pi / 3.0;
const double I = (3.0 / std::numbers::pi) * std::log(2.0 + std::sqrt(3.0)) - 0.96;
const double K = 1.0 / I;
const double LN_2_PLUS_SQRT_3 = std::log(2.0 + std::sqrt(3.0));

std::mt19937_64 &getRDMengine()
{
    thread_local std::mt19937_64 engine(std::random_device{}());
    return engine;
}

double my_round(double x, int p)
{
    double powOf10 = std::pow(10.0, p);
    double val = std::floor(x * powOf10) / powOf10;

    return std::round(val * 1000.0) / 1000.0;
}

Rounder::Rounder(int p)
    : precision_{p}
{
}

std::pair<double, double> Rounder::operator()(double x, double y)
{
    double newX = my_round(dx_ + x, precision_);
    double newY = my_round(dy_ + y, precision_);
    dx_ += x - newX;
    dy_ += y - newY;
    return {newX, newY};
}

double my_sin(double degree)
{
    return std::sin(degree * std::numbers::pi / 180.0);
}

double my_cos(double degree)
{
    return std::cos(degree * std::numbers::pi / 180.0);
}

double norm2(const std::array<double, 2> &v)
{
    return std::hypot(v[0], v[1]);
}

double dist2(const std::array<double, 2> &a, const std::array<double, 2> &b)
{
    return std::hypot(a[0] - b[0], a[1] - b[1]);
}

double normInf(const std::array<double, 2> &v)
{
    return std::max(std::abs(v[0]), std::abs(v[1]));
}

double distInf(const std::array<double, 2> &a, const std::array<double, 2> &b)
{
    return std::max(std::abs(a[0] - b[0]), std::abs(a[1] - b[1]));
}

int randint(int a, int b)
{
    std::uniform_int_distribution<int> dist(a, b);
    return dist(getRDMengine());
}

int randint(int max)
{
    return randint(0, max);
}

double f(double x)
{
    return K * (1.0 / (a * (x - 0.5)) - 0.96);
}

double F(double x)
{
    double u = a * (x - 0.5);
    return K * ((1.0 / a) * (std::log(1.0 / std::cos(u) + std::tan(u)) + LN_2_PLUS_SQRT_3) - 0.96 * x);
}

double my_random()
{
    std::uniform_real_distribution<double> dist(0.0, 1.0);
    double u = dist(getRDMengine());
    double x = u;
    constexpr double eps = 1e-9;

    for (int i = 0; i < 6; ++i)
    {
        if (std::abs(x - 0.5) < eps)
        {
            x += eps;
        }

        double fx = f(x);
        if (fx != 0.0)
        {
            x -= (F(x) - u) / fx;
        }

        x = std::clamp(x, eps, 1.0 - eps);
    }
    return x;
}

double round(double x, size_t precision)
{
    size_t pow10 = std::pow(10, precision);
    return floor(x * pow10) / pow10;
}